R/get_error_path.R

```diff
@@ -70,23 +103,35 @@ get_error_path <- function(schema, element = "target_metadata",
     path <- paste(path, "items", sep = "/")
   }
 
-
   switch(type,
     schema = if (length(path) == 0L) NA else paste0("#", path),
     instance = generate_instance_path_glue(path)
   )
 }
 
+# To create instance path, remove `properties/` elements and substitute `items`
+# elements with error position index glue interpolation expression. to be evaluated
+# later via glue::glue_data().
 generate_instance_path_glue <- function(path) {
+  item_props <- c(
+    "{round_i - 1}",
+    "{model_task_i - 1}",
+    "{target_key_i - 1}"
+  )
   split_path <- gsub("properties/", "", path) %>%
     strsplit("/") %>%
     unlist()
-
   is_item <- split_path == "items"
-  split_path[is_item] <- c(
-    "{round_i - 1}",
-    "{model_task_i - 1}",
-    "{target_key_i - 1}"
-  )[1:sum(is_item)]
+  n_item_props <- sum(is_item)
+  split_path[is_item] <- utils::head(item_props, n_item_props)
```

Comment:
```suggestion
  split_path[is_item] <- item_props[seq_len(n_item_props)]
```

subsetting feels more natural and is faster by an order of magnitude.

``` r
x <- 1:10
bench::mark(utils::head(x, 3), x[seq_len(3)])
#> # A tibble: 2 × 6
#>   expression             min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>        <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 utils::head(x, 3)   1.52µs   1.68µs   511741.        0B     51.2
#> 2 x[seq_len(3)]         41ns     82ns  8051712.        0B      0
```

<sup>Created on 2024-11-06 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
These really are tiny performance improvements when we're talking about a length 3 vector.
More importantly, indexing returns an error when `n_item_props` is 0L while `head()` does not
``` r
1:3[seq_len(0L)]
#> Error in 1:3[seq_len(0L)]: argument of length 0

head(1:3, 0L)
#> integer(0)
```
I also find `head()` much clearer

Reply:
Actually the above is a problem with subsetting a `:` expression. Subsetting with a zero length works for already defined objects, so I've gone ahead and changed it.

``` r
paths <- vector(length = 3L)
n <- sum(paths)
paths[seq_len(n)]
#> logical(0)
```

<sup>Created on 2024-11-07 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>